### PR TITLE
Generic patching of bundled manifests #70

### PR DESCRIPTION
**Pull request by sainivikas**
-p : path of patches

### CODE REVIEW

**Comment by mazzy89** [comment]
@sainivikas have you explored here the possibility to consume `kustomize` as a library and hence avoid installation of external dependency?

**Comment by sainivikas** [comment]
> @sainivikas have you explored here the possibility to consume `kustomize` as a library and hence avoid installation of external dependency?

@mazzy89 `kustomize` has been added as library

File: Makefile [hunk 1]
```diff
@@ -10,7 +10,8 @@ endif
 setup:
 	which esc 2>&1 > /dev/null || go get -u github.com/mjibson/esc
 	which github-release 2>&1 > /dev/null || go get github.com/aktau/github-release
-
+	curl -s "https://raw.githubusercontent.com/\
+kubernetes-sigs/kustomize/master/hack/install_kustomize.sh"  | bash
```
Comment (mazzy89) [inline Makefile]: is this still needed here?

File: cmd/common.go [hunk 2]
```diff
@@ -77,6 +77,13 @@ func getConfig(cmd *cobra.Command) types.PlatformConfig {
 		log.Infof("Running a dry-run mode, no changes will be made")
 	}
 
+	patchFilePath, _ := cmd.Flags().GetString("path")
+	base.PatchPath = patchFilePath
```
Comment (mazzy89) [inline cmd/common.go]: you assign `patchFilePath` to `base.PatchPath` and then check `patchFilePath` if it is not empty. Could you explain why?

File: pkg/platform/platform.go [hunk 3]
```diff
@@ -513,6 +515,63 @@ func (platform *Platform) WaitForNamespace(ns string, timeout time.Duration) {
...
+		log.Fatal(err)
+	}
+}
+
+func GetFinalPatchYaml(patchFilePath string) string {
+	kustWorkingDir := "overlays/patchDir/"
+	os.RemoveAll(kustWorkingDir)
+	err := os.MkdirAll(kustWorkingDir, 0755)
+   	if err != nil {
+    	log.Fatal(err)
```
Comment (mazzy89) [inline pkg/platform/platform.go]: Could you run your IDE, text editor with go tools and follow the best practices? the function has spaces that are not accepted in golang. Please use `gofmt` and run automatically.

File: pkg/platform/platform.go [hunk 4]
```diff
@@ -524,16 +583,37 @@ func (platform *Platform) ApplySpecs(namespace string, specs ...string) error {
 			if err != nil {
 				return err
 			}
-			if err := kubectl("apply %s -f %s", namespace, dir); err != nil {
-				return err
+			if(platform.PatchPath != "") {
+				baseYamlContent, err := ioutil.ReadFile(dir)
+				if err != nil {
+			        log.Fatal(err)
+			    }
```
Comment (mazzy89) [inline pkg/platform/platform.go]: Same here `gofmt`.

File: pkg/platform/platform.go [hunk 5]
```diff
@@ -524,16 +583,37 @@ func (platform *Platform) ApplySpecs(namespace string, specs ...string) error {
...
 			}
 		} else {
 			template, err := platform.Template(spec, "manifests")
 			if err != nil {
 				return err
 			}
-			if err := kubectl("apply %s -f %s", namespace, text.ToFile(template, ".yaml")); err != nil {
-				return err
+			if(platform.PatchPath != "")
+			{
```
Comment (mazzy89) [inline pkg/platform/platform.go]: Same here `gofmt`.

File: overlays/overlays_kustomization_template [hunk 6]
```diff
@@ -0,0 +1,8 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+
+bases:
+  - ../baseDir/
+
+patchesStrategicMerge:
+%s
```
Comment (mazzy89) [inline overlays/overlays_kustomization_template]: Could you leverage here Go Template? https://golang.org/pkg/text/template/ Template is a powerful language with a lot of primitives pretty robust to manipulate a file. This way is error-prone.

File: pkg/platform/platform.go [hunk 7]
```diff
@@ -513,6 +515,63 @@ func (platform *Platform) WaitForNamespace(ns string, timeout time.Duration) {
...
+	os.RemoveAll(kustWorkingDir)
+	err := os.MkdirAll(kustWorkingDir, 0755)
+   	if err != nil {
+    	log.Fatal(err)
+    }
+	files, _ := ioutil.ReadDir(patchFilePath)
+	var buffer bytes.Buffer
+	for _, file := range files {
+	    if(strings.HasSuffix(file.Name(), ".yaml") || strings.HasSuffix(file.Name(), ".yml")){
+	    	buffer.WriteString("  - ")
```
Comment (mazzy89) [inline pkg/platform/platform.go]: You would get rid off this if you would leverage Go Template here and reading and writing on a file would be easier.

File: Makefile (hunk 1 above)
Comment (sainivikas) [reply Makefile]: not needed, removed.

File: cmd/common.go (hunk 2 above)
Comment (sainivikas) [reply cmd/common.go]: no specific reason, can assign to base.PatchPath directly. changes done.

File: overlays/overlays_kustomization_template (hunk 6 above)
Comment (sainivikas) [reply overlays/overlays_kustomization_template]: sure, I will check and will push the respective change.

File: pkg/platform/platform.go (hunk 5 above)
Comment (sainivikas) [reply pkg/platform/platform.go]: Done

File: pkg/platform/platform.go (hunk 4 above)
Comment (sainivikas) [reply pkg/platform/platform.go]: Done

File: pkg/platform/platform.go (hunk 3 above)
Comment (sainivikas) [reply pkg/platform/platform.go]: sure, done formatting.

File: cmd/common.go [hunk 8]
```diff
@@ -77,10 +77,9 @@ func getConfig(cmd *cobra.Command) types.PlatformConfig {
 		log.Infof("Running a dry-run mode, no changes will be made")
 	}
 
-	patchFilePath, _ := cmd.Flags().GetString("path")
-	base.PatchPath = patchFilePath
-	if (patchFilePath != ""){
-		_ , err := ioutil.ReadDir(patchFilePath)
+	base.PatchPath, _ := cmd.Flags().GetString("path")
+	if (base.PatchPath != ""){
```
Comment (mazzy89) [inline cmd/common.go]: Please remove brackets from here.

File: pkg/platform/platform.go [hunk 9]
```diff
@@ -513,6 +515,63 @@ func (platform *Platform) WaitForNamespace(ns string, timeout time.Duration) {
...
+	os.RemoveAll(kustWorkingDir)
+	err := os.MkdirAll(kustWorkingDir, 0755)
+	if err != nil {
+		log.Fatal(err)
+	}
+	files, _ := ioutil.ReadDir(patchFilePath)
+	var buffer bytes.Buffer
+	for _, file := range files {
+		if strings.HasSuffix(file.Name(), ".yaml") || strings.HasSuffix(file.Name(), ".yml") {
+			buffer.WriteString("  - ")
```
Comment (mazzy89) [inline pkg/platform/platform.go]: As discussed in the previous review, here would be better to leverage Go template. Any obstacles that do not the implementation here?

Comment (sainivikas) [reply pkg/platform/platform.go]: @mazzy89 on it.

File: cmd/common.go (hunk 8 above)
Comment (sainivikas) [reply cmd/common.go]: Done

File: pkg/platform/platform.go (hunk 9 above)
Comment (sainivikas) [reply pkg/platform/platform.go]: @mazzy89 Using Go template now.

**Pull request closed without merging**